shielddb-core/src/test/java/nf/fr/k49/shielddb/core/ShieldDBBuilderTest.java

```diff
@@ -0,0 +1,212 @@
+package nf.fr.k49.shielddb.core;
+
+import nf.fr.k49.shielddb.core.shield.BottomShield;
+import nf.fr.k49.shielddb.core.shield.ShieldDBShield;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Test;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class ShieldDBBuilderTest {
+
+    @Test
+    public void firstShieldPlacedAtBeginningOfShield() {
+        ShieldDBShield<String> anyFirstShield = new MockDBShield();
+        ShieldDBShield<String> anySecondShield = new MockDBShield();
+
+        List<String> build = ShieldDB.<String>builder()
+                .shield(anyFirstShield)
+                .shield(anySecondShield)
+                .build();
+
+        MatcherAssert.assertThat(build, Matchers.is(anyFirstShield));
+    }
+
+    @Test
+    public void shieldDbWithoutShieldsHasOnlyBottomShield() {
+        List<String> build = ShieldDB.<String>builder()
+                .build();
+
+        MatcherAssert.assertThat(build, Matchers.instanceOf(BottomShield.class));
+    }
+
+    @Test
+    public void builderNestsShields() {
+        ShieldDBShield<String> anyFirstShield = new MockDBShield();
+        ShieldDBShield<String> anySecondShield = new MockDBShield();
+
+        List<String> build = ShieldDB.<String>builder()
+                .shield(anyFirstShield)
+                .shield(anySecondShield)
+                .build();
+
+        MockDBShield shield = (MockDBShield)build;
+        MatcherAssert.assertThat(shield.getNextShield(), Matchers.is(anySecondShield));
+    }
+
+    @Test
+    public void lastElementIsBottomShield() {
+        ShieldDBShield<String> anyFirstShield = new MockDBShield();
+        ShieldDBShield<String> anySecondShield = new MockDBShield();
+
+        List<String> build = ShieldDB.<String>builder()
+                .shield(anyFirstShield)
+                .shield(anySecondShield)
+                .build();
+
+        MockDBShield shield = (MockDBShield)build;
+        MatcherAssert.assertThat(shield.getNextShield().getNextShield(), Matchers.instanceOf(BottomShield.class));
+    }
+
+    @Test
+    public void lastElementOfManyConfiguredShieldIsBottomShield() {
+        int countShields = 20;
+        List<MockDBShield> shields = IntStream.range(0, countShields).mapToObj((i) -> new MockDBShield()).collect(Collectors.toList());
+
+        ShieldDBBuilder<String> builder = ShieldDB.<String>builder();
+        shields.stream().forEach(builder::shield);
+
+        ShieldDBShield<String> db = (ShieldDBShield<String>)builder.build();
+
+        for (int i = 0; i < countShields; i++) {
+            db = db.getNextShield();
+        }
+        MatcherAssert.assertThat(db, Matchers.instanceOf(BottomShield.class));
+    }
+
+
+    private static class MockDBShield implements ShieldDBShield<String> {
```

Comment:
You can implements PassThroughShield (https://github.com/kuroidoruido/ShieldDB/blob/master/shielddb-core/src/main/java/nf/fr/k49/shielddb/core/shield/PassThroughShield.java) instead of ShieldDBShield. This way you can avoid implementing a lot of methods.

Reply:
Aaah, I don't get the complete survey of all classes ;)
My additional plan is using a really mock (e.g. via mockito) so that a test can check calling some methods like setNextShield etc. Would this a better way? If so, I change the test by using the mock. Alternatively I can use the PassThroughShield. Give me a hint ;)